[Jared Wong] iP

Jajared opened the pull request:
# <img src="https://media.giphy.com/media/WUlplcMpOCEmTGBtBW/giphy.gif" width="40"> GeePeeTee: Your Java Task Manager ✨ 
> The only Java I love is Java Chip Frappuccino ☕  - Jared

Manage your tasks efficiently with **GeePeeTee**, the Java-based task manager that keeps you organized and productive.

## Features:

- **Ease of Use**: Simply intuitive
- **Cross-Platform**: Works on multiple platforms
- **Open Source**: Dive into the code

## Upcoming Features:
- [ ] Add JavaFx GUI
- [ ] :tada: Celebrate the first release!

> [!IMPORTANT]
> ## How to Get Started
1. Clone the repository:
```bash
git clone https://github.com/Jajared/ip.git
```
2. Navigate to the project directory
```bash
cd src/main/java
```
3. Compile the source code and run the application
```bash
javac Main.java && java Main
```

This project was forked from [CS2103T ip Repo](https://github.com/nus-cs2103-AY2324S2/ip/). Many thanks to the awesome CS2103T Teaching Team! 👍 🏅 💯

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Jajared committed:
refactor: changed message and updated git ignore file

Jajared committed:
chore: updated git ignore file to exclude class files

Jajared committed:
feat: added user input

Jajared committed:
feat: added task list

Jajared committed:
feat: added mark and list functionality

Jajared committed:
feat: added event, deadline and todo

Jajared committed:
test: added simple automated testing

Jajared committed:
refactor: added error handling

Jajared committed:
feat: added delete task

Jajared committed:
feat: add file save functionality

Jajared committed:
refactor: change to use date object for tasks

Jajared committed:
Merge branch 'branch-Level-8'

Jajared committed:
refactor: refactor code to include more OOP classes.

Jajared committed:
refactor: Add ui class to deal with user interaction.

Jajared committed:
refactor: package classes

Jajared committed:
Merge branch 'branch-A-MoreOOP'

Jajared committed:
refactor: add new custom exception classes

Jajared committed:
refactor: Remove unnecessary imports.

Jajared committed:
Merge branch 'branch-A-Packages'

Jajared committed:
Merge pull request #1 from Jajared/add-gradle-support

Add gradle support

Jajared committed:
build: Add gradle support.

Jajared committed:
Merge branch 'branch-A-Gradle'

Jajared committed:
test: Add JUnit test cases for Ui and Event classes.

Jajared committed:
Merge branch 'branch-A-JUnit'

Jajared committed:
build: Update gradle to build jar file

Jajared committed:
Merge branch 'branch-A-Jar'

Jajared committed:
docs: Add javadoc

Jajared committed:
style: Change code standards

Jajared committed:
Merge branch 'branch-A-CodingStandard'

Jajared committed:
Merge branch 'branch-A-JavaDoc'

Jajared committed:
style: Add white spaces

Jajared committed:
feat: Add find functionality

Jajared committed:
Merge branch 'branch-Level-9'

Jajared committed:
style: Fix styling and formatting issues

Jajared committed:
docs: Update README.md

Jajared committed:
style: Add checkstyle config files and fix style errors

Jajared committed:
style: Fix javadocs styling

Jajared committed:
Merge branch 'branch-A-CheckStyle'

Jajared committed:
feat: Add gui framework

UdhayaShan1 commented on README.md:
```diff
@@ -1,3 +1,35 @@
+# <img src="https://media.giphy.com/media/WUlplcMpOCEmTGBtBW/giphy.gif" width="40"> GeePeeTee: Your Java Task Manager ✨ 
+> The only Java I love is Java Chip Frappuccino ☕  - Jared
+
+Manage your tasks efficiently with **GeePeeTee**, the Java-based task manager that keeps you organized and productive.
```
Good readme!

UdhayaShan1 commented on src/main/java/GeePeeTee.java:
```diff
@@ -0,0 +1,65 @@
+            storage = new Storage("GeePeeTee.txt");
+            taskList = new TaskList(storage.loadTaskList());
+        } catch (FileNotFoundException e) {
+            ui.showFileNotFoundError();
+        } catch (IOException e) {
+            ui.showLoadingError();
+        } catch (GeePeeTeeException e) {
+            ui.showErrorMessage(e.getMessage());
```
Good use of functions to print

UdhayaShan1 commented on src/main/java/GeePeeTee.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class GeePeeTee {
+
+    private TaskList taskList;
```
```suggestion
    private TaskList listOfTasks;
```

UdhayaShan1 commented on src/main/java/GeePeeTee.java:
```diff
@@ -0,0 +1,65 @@
+        Parser parser = new Parser(taskList, storage, ui);
+        while (!input.equals("bye")) {
+            input = scanner.nextLine();
+            System.out.println("\n--------------------------------------------------");
```
Perhaps can abstract out the line to ui helper method as well.

UdhayaShan1 commented on src/main/java/exception/GeePeeTeeException.java:
```diff
@@ -0,0 +1,31 @@
+ * with custom messages.
+ * </p>
+ */
+public class GeePeeTeeException extends Exception {
```
Many custom exception classes is good.

UdhayaShan1 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,272 @@
+            break;
+        case "delete":
+            processDeleteCommand(input);
+            break;
```
Good use of switch, cleaner than if-else control flow.

UdhayaShan1 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
Perhaps a clearer variable name for the date

UdhayaShan1 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,78 @@
+            return new Event(description, parsedAt, parsedBy);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new InvalidTaskFormatException(
+                    "Invalid event format. Please use 'event description /from date /to date'.");
```
Good use of exceptions here

UdhayaShan1 commented on src/test/java/ui/UiTest.java:
```diff
@@ -0,0 +1,67 @@
+     * The error message is displayed correctly.
+     */
+    @Test
+    public void showErrorMessage_correctOutput() {
```
Good tests

UdhayaShan1 reviewed (commented):
Overall good!

shayaansultan commented on src/main/java/GeePeeTee.java:
```diff
@@ -0,0 +1,65 @@
+        try {
+            storage = new Storage("GeePeeTee.txt");
+            taskList = new TaskList(storage.loadTaskList());
+        } catch (FileNotFoundException e) {
```
Good use of try catch to account for different types of errors 👍

shayaansultan commented on src/main/java/exception/InvalidTaskIndexException.java:
```diff
@@ -0,0 +1,36 @@
+ * errors occur.
+ * </p>
+ */
+public class InvalidTaskIndexException extends Exception {
```
Good use of multiple exception classes for different scenarios

shayaansultan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,272 @@
+                throw new InvalidTaskIndexException("The index of a task cannot be 0 or negative.");
+            }
+            Task taskToMark = taskList.getTask(markIndex);
+            if (taskToMark == null) {
```
Good error handling to deal with bad inputs

shayaansultan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,272 @@
+     * @throws GeePeeTeeException         If an error occurs while adding the todo
+     *                                    task.
+     */
+    private void processToDoCommand(String input) {
```
Maybe divide the Parser class further into Parser and Handler/Processor class to streamline the code and improve readability.

shayaansultan commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+    public static Deadline createFromInput(String input) throws InvalidTaskFormatException, InvalidDateException {
+        try {
+            String[] parts = input.split(" /by ");
+            String description = parts[0].split("deadline ")[1];
```
Maybe let the Parser class handle parsing the raw input

shayaansultan commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,141 @@
+        showTaskMessage(task);
+    }
+
+    /**
```
Good JavaDoc Headers throughout the project 👍

shayaansultan reviewed (commented):
LGTM!

Jajared committed:
feat: Add javafx gui

Jajared committed:
test: Fix test cases and minor refactoring

Jajared committed:
Merge branch 'branch-Level-10'

Jajared committed:
refactor: Refactor create task handling to parser and small ui changes

Jajared committed:
Add assertion to validate assumptions in project

- Introduce assertions in the Main class to verify the successful
  loading of FXML files and controller injection
- Introduce assertions in storage class to verify proper file handling
- Introduce assertions in GeePeeTee class to verify successful
  instantiation of key components

The addition of these assertions aims to catch potential errors early in
the cycle, improving code reliability and maintainability.

This update does not alter any existing functionalities.

Jajared committed:
Refactor to improve code quality

- Implemented early input validation in parser class to minimise dead
  code and improve input handling
- Refactored some code to adhere to abstraction level standards

Overall, this commit represents an effort to improve the code quality.

Jajared committed:
Refactor to use Java Streams

- Implement stream for TaskList
- Implement stream for Parser
- Add date handling in Deadline and Event
This refactor by using streams allows processing of collections of data
in a more efficient way without altering its functionality.

Jajared committed:
Merge pull request #3 from Jajared/branch-A-CodeQuality

Refactor to improve code quality

Jajared committed:
Merge with branch-A-Streams

Jajared committed:
Fix date handling in parser

Jajared committed:
Merge with branch-A-Assertions

Jajared committed:
feat: Add priority system to tasks

Jajared committed:
feat: Improve error handling message and minor bug fixes

Jajared committed:
Merge branch-C-Priority

Jajared committed:
refactor: Improve ui

Jajared committed:
docs: Add more java docs and refactor parser

Jajared committed:
Merge branch 'branch-A-Personality'

Jajared committed:
test: Add extra junit tests

Jajared committed:
Merge branch 'branch-A-MoreTesting'

Jajared committed:
docs: Add product showecase image

Jajared committed:
refactor: Add extra error handling for tag command

Jajared committed:
Merge branch 'branch-A-MoreErrorHandling'

Jajared committed:
docs: Update readme

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
docs: Update readme and minor bug fixes

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
Update README.md

Jajared committed:
Update README.md by adding FAQ section

Jajared committed:
Update README.md

Jajared committed:
Merge branch 'branch-A-UserGuide'

Jajared committed:
docs: Add release link to readme

Jajared committed:
docs: Update jar file naming

Jajared committed:
Merge branch 'branch-A-Release'

Jajared committed:
docs: Update minor typo
